grid/fetch-demo

sebastianbochan opened the pull request:
Added fetch-data Grid demo.

stitot reviewed (changes requested):
I like this, but have one rather big objection, so let's deal with that before the full review:

Even thought rendering of `this.value.RootDisk.UsedGB` etc. works fine our support for objects as values is limited. Breaks sorting (ref. https://github.com/highcharts/highcharts/issues/22936) and isn't editable. Instead I suggest the approach below (courtesy of ChatGPT, so beware). Then follow my suggestion of hiding column in https://github.com/highcharts/highcharts/issues/22936 instead.

This way the column is at least sortable by `UsedGB` and that value can be edited. Still a shortcoming that its not sortable by `SizeGB` and that `SizeGB` can not be edited, but it's still better IMO.

```
function parseFields(items, paths) {
  return paths.reduce((acc, path) => {
    const keys = path.split('.');
    acc[path] = items.map(item =>
      keys.reduce((obj, key) => obj?.[key], item)
    );
    return acc;
  }, {});
}

async function fetchAndParse(requestedProps) {
  const resp = await fetch(
    'https://demo-live-data.highcharts.com/instances.json'
  );
  if (!resp.ok) {
    throw new Error(`HTTP ${resp.status}`);
  }
  const instances = await resp.json();
  return parseFields(instances, requestedProps);
}

const fields = [
  "InstanceId",
  "InstanceType",
  "DiskSpace.RootDisk.SizeGB",
  "DiskSpace.RootDisk.UsedGB",
  ...
];

fetchAndParse(props)
  .then(data => {
    initGrid(data);
  })
  .catch(console.error);

```

karolkolodziej commented on samples/grid-lite/demo/parse-data/demo.css:
```diff
@@ -0,0 +1,24 @@
+}
+
+.hcg-table td[data-column-id="HealthIndicator"] {
+    text-align: center;
```
It looks a bit off when it's centred 🤔

sebastianbochan replied:
Currently we have whe width option, so will adjust it.

karolkolodziej commented on samples/grid-lite/demo/parse-data/demo.details:
```diff
@@ -0,0 +1,12 @@
+---
+name: Fetch Data Grid
```
```suggestion
name: Grid with data from an external source
```

KamilKubik reviewed (approved):
Shouldn't we rename the folder name to `external-source-data` or something?

sebastianbochan commented:
@stitot 
1. I modified my parser a bit; the sorting option works. In between, I would rather use the simplest parser as it can (instead of using the `reduce` function, suggested by our close friend GPT) because of our users' different entry levels. IMO, the demo should be the easiest to understand.
2. Refactored columns / headers a bit to hide column.

ddragula reviewed (approved):
Looks good!

stitot commented on samples/grid-lite/demo/parse-data/demo.css:
```diff
@@ -0,0 +1,25 @@
+    padding: 0 20px;
+}
+
+.hcg-table td[data-column-id="HealthIndicator"],
```
Suggestion: instead use `.centered` and add that class name via the `className` option. A more generic and reusable approach, and what we recommend instead of targeting individual columns in CSS.

stitot commented on samples/grid-lite/demo/parse-data/demo.html:
```diff
@@ -0,0 +1,6 @@
+
+<div id="container"></div>
+<p class="highcharts-description">
+    Grid with dynamically loaded data from the external resource.
```
This demo show how you can fetch JSON from an external source and then shape that dataset into a format that Grid can consume.

stitot commented on samples/grid-lite/demo/parse-data/demo.js:
```diff
@@ -0,0 +1,100 @@
+            columns: data
+        },
+        rendering: {
+            columns: {
+                distribution: 'mixed'
+            }
```
```suggestion
            rows: {
                strictHeights: true
            }
```
columns.distribution is not needed, (turned on automatically if `width` exists) and potentially adds confusion. But `strictHeights` is nice, to prevent some ugly line/word breaks.

stitot commented on samples/grid-lite/demo/parse-data/demo.js:
```diff
@@ -0,0 +1,100 @@
+            {
+                id: 'HealthIndicator',
+                width: '100px',
+                useHTML: true,
```
This is not an option in Grid, and not needed

stitot commented on samples/grid-lite/demo/parse-data/demo.js:
```diff
@@ -0,0 +1,100 @@
+                    formatter: function () {
+                        const val = this.value;
+                        return `<img
+                            src='https://www.highcharts.com/samples/graphics/dashboards/cloud-monitoring/${val.toLowerCase()}-ico.${val === 'Critical' ? 'png' : 'svg'}'
```
This is just confusing, IMO. Not possible to get hold of `critical.svg` somehow?

stitot commented on samples/grid-lite/demo/parse-data/demo.js:
```diff
@@ -0,0 +1,100 @@
+(async () => {
+    const data = {};
+
+    // Load data
```
```suggestion
    // Fetch JSON data from external source
```

stitot commented on samples/grid-lite/demo/parse-data/demo.js:
```diff
@@ -0,0 +1,100 @@
+        'https://demo-live-data.highcharts.com/instances.json'
+    ).then(response => response.json());
+
+    // Fields that we would like to extract from data
```
```suggestion
    // The properties we would like to extract from the JSON
```

stitot commented on samples/grid-lite/demo/parse-data/demo.js:
```diff
@@ -0,0 +1,100 @@
+        'HealthIndicator'
+    ];
+
+    // Parse data
```
```suggestion
    // Parse the properties to a data format that Grid can consume (https://www.highcharts.com/docs/grid/understanding-grid#datatable)
```

stitot reviewed (changes requested):
Very nice! Just a few inline comments.
